nexus/saga-recovery/src/lib.rs

```diff
@@ -0,0 +1,659 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Saga recovery bookkeeping
+//!
+//! If you're reading this, you first want to read the big block comment in the
+//! saga recovery background task.  It explains important background about what
+//! we're trying to do here.  The rest of this comment assumes you've read all
+//! that.
+//!
+//! ## Saga recovery passes
+//!
+//! For the reasons mentioned in that comment, saga recovery is done in
+//! **passes**.  The state that's kept between passes (when the task is "at
+//! rest") is called the [`RestState`].
+//!
+//! Each saga recovery pass looks like this:
+//!
+//! 1. Start with initial [`RestState`] and [`Report`].
+//! 2. List in-progress sagas from the database.
+//! 3. Collect list of sagas that have been started by the rest of Nexus.
+//! 4. Use [`Plan::new()`] to construct a plan.  The plan examines all the sagas
+//!    reported by the database as well as all the sagas we knew about before
+//!    and determines for each one exactly what's needed.  Each saga falls into
+//!    one of a few buckets:
+//!    * It's clearly not running, but should be, so it needs to be recovered.
+//!    * It's clearly running, so it does not need to be recovered.
+//!    * It has finished running and we can stop keeping track of it.
+//!    * It _may_ be running but we cannot tell because of the intrinsic race
+//!      between steps 2 and 3 above.  We'll keep track of these and resolve the
+//!      ambiguity on the next pass.
+//! 5. Carry out recovery for whatever sagas need to be recovered.  Use
+//!    [`ExecutionBuilder::new()`] to construct a description of what happened.
+//! 6. Update the [`RestState`] and [`Report`] to reflect what happened in this
+//!    pass.
```

Comment:
just wanted to say that this comment --- and the rest of the docs in this module --- are really lovely, thank you for writing this all down <3